File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/metrics/NavigationEventFactory.java
package com.mapbox.services.android.navigation.v5.navigation.metrics;

import android.location.Location;
import android.support.annotation.NonNull;

import com.mapbox.android.telemetry.TelemetryUtils;
import com.mapbox.services.android.navigation.v5.routeprogress.MetricsRouteProgress;
import com.mapbox.services.android.navigation.v5.utils.DistanceFormatter;

import java.util.Date;
import java.util.List;

public class NavigationEventFactory {
  static final int EVENT_VERSION = 7;

  private NavigationEventFactory() {

  }

  public static NavigationDepartEvent buildNavigationDepartEvent(@NonNull PhoneState phoneState,
                                                                 @NonNull SessionState sessionState,
                                                                 @NonNull MetricsRouteProgress metricProgress,
                                                                 @NonNull Location location,
                                                                 @NonNull String sdkIdentifier) {
    NavigationDepartEvent navigationDepartEvent = new NavigationDepartEvent(phoneState);
    setEvent(sessionState, metricProgress, location, sdkIdentifier, navigationDepartEvent);
    return navigationDepartEvent;
  }

  public static NavigationCancelEvent buildNavigationCancelEvent(@NonNull PhoneState phoneState,
                                                                 @NonNull SessionState sessionState,
                                                                 @NonNull MetricsRouteProgress metricProgress,
                                                                 @NonNull Location location,
                                                                 @NonNull String sdkIdentifier) {

    NavigationCancelEvent navigationCancelEvent = new NavigationCancelEvent(phoneState);
    setEvent(sessionState, metricProgress, location, sdkIdentifier, navigationCancelEvent);
    String arrivalTimestamp = TelemetryUtils.generateCreateDateFormatted(sessionState.arrivalTimestamp());
    navigationCancelEvent.setArrivalTimestamp(arrivalTimestamp);
    return navigationCancelEvent;
  }

  public static NavigationArriveEvent buildNavigationArriveEvent(@NonNull PhoneState phoneState,
                                                                 @NonNull SessionState sessionState,
                                                                 @NonNull MetricsRouteProgress metricProgress,
                                                                 @NonNull Location location,
                                                                 @NonNull String sdkIdentifier) {
    NavigationArriveEvent navigationArriveEvent = new NavigationArriveEvent(phoneState);
    setEvent(sessionState, metricProgress, location, sdkIdentifier, navigationArriveEvent);
    return navigationArriveEvent;
  }


  public static NavigationRerouteEvent buildNavigationRerouteEvent(@NonNull PhoneState phoneState,
                                                                   @NonNull SessionState sessionState,
                                                                   @NonNull MetricsRouteProgress metricProgress,
                                                                   @NonNull Location location,
                                                                   @NonNull String sdkIdentifier,
                                                                   @NonNull RerouteEvent rerouteEvent) {
    NavigationRerouteEvent navigationRerouteEvent =
      new NavigationRerouteEvent(phoneState, rerouteEvent, metricProgress);
    setEvent(sessionState, metricProgress, location, sdkIdentifier, navigationRerouteEvent);
    List<Location> locationList = sessionState.beforeEventLocations();
    navigationRerouteEvent.setLocationsBefore(convertToArray(locationList));
    navigationRerouteEvent.setLocationsAfter(convertToArray(sessionState.afterEventLocations()));
    navigationRerouteEvent.setSecondsSinceLastReroute(sessionState.secondsSinceLastReroute());
    return navigationRerouteEvent;
  }


  public static NavigationFeedbackEvent buildNavigationFeedbackEvent(@NonNull PhoneState phoneState,
                                                                     @NonNull SessionState sessionState,
                                                                     @NonNull MetricsRouteProgress metricProgress,
                                                                     @NonNull Location location,
                                                                     @NonNull String sdkIdentifier,
                                                                     String description, String feedbackType,
                                                                     String screenshot, String feedbackSource) {
    NavigationFeedbackEvent navigationFeedbackEvent = new NavigationFeedbackEvent(phoneState, metricProgress);
    setEvent(sessionState, metricProgress, location, sdkIdentifier, navigationFeedbackEvent);
    navigationFeedbackEvent.setLocationsBefore(convertToArray(sessionState.beforeEventLocations()));
    navigationFeedbackEvent.setLocationsAfter(convertToArray(sessionState.afterEventLocations()));
    navigationFeedbackEvent.setDescription(description);
    navigationFeedbackEvent.setFeedbackType(feedbackType);
    navigationFeedbackEvent.setScreenshot(screenshot);
    navigationFeedbackEvent.setSource(feedbackSource);
    return navigationFeedbackEvent;
  }

  private static void setEvent(SessionState sessionState, MetricsRouteProgress metricProgress, Location location,
                               String sdkIdentifier, NavigationEvent navigationEvent) {
    navigationEvent
      .setAbsoluteDistanceToDestination(DistanceFormatter.calculateAbsoluteDistance(location, metricProgress));
    navigationEvent
      .setDistanceCompleted((int) (sessionState.eventRouteDistanceCompleted() + metricProgress.getDistanceTraveled()));
    navigationEvent.setDistanceRemaining(metricProgress.getDistanceRemaining());
    navigationEvent.setDurationRemaining(metricProgress.getDurationRemaining());
    navigationEvent.setProfile(metricProgress.getDirectionsRouteProfile());
    navigationEvent.setLegIndex(metricProgress.getLegIndex());
    navigationEvent.setLegCount(metricProgress.getLegCount());
    navigationEvent.setStepIndex(metricProgress.getStepIndex());
    navigationEvent.setStepCount(metricProgress.getStepCount());
    navigationEvent.setEstimatedDistance(metricProgress.getDirectionsRouteDistance());
    navigationEvent.setEstimatedDuration(metricProgress.getDirectionsRouteDuration());
    navigationEvent.setStartTimestamp(obtainStartTimestamp(sessionState));
    navigationEvent.setEventVersion(EVENT_VERSION);
    navigationEvent.setSdkIdentifier(sdkIdentifier);
    navigationEvent.setSessionIdentifier(sessionState.sessionIdentifier());
    navigationEvent.setLat(location.getLatitude());
    navigationEvent.setLng(location.getLongitude());
    navigationEvent.setGeometry(sessionState.currentGeometry());
    navigationEvent.setSimulation(sessionState.mockLocation());
    navigationEvent.setLocationEngine(sessionState.locationEngineName());
    navigationEvent.setTripIdentifier(sessionState.tripIdentifier());
    navigationEvent.setRerouteCount(sessionState.rerouteCount());
    navigationEvent.setOriginalRequestIdentifier(sessionState.originalRequestIdentifier());
    navigationEvent.setRequestIdentifier(sessionState.requestIdentifier());
    navigationEvent.setOriginalGeometry(sessionState.originalGeometry());
    navigationEvent.setOriginalEstimatedDistance(sessionState.originalDistance());
    navigationEvent.setOriginalEstimatedDuration(sessionState.originalDuration());
    navigationEvent.setOriginalStepCount(sessionState.originalStepCount());
    navigationEvent.setPercentTimeInForeground(sessionState.percentInForeground());
    navigationEvent.setPercentTimeInPortrait(sessionState.percentInPortrait());
    navigationEvent.setTotalStepCount(sessionState.currentStepCount());
  }

  private static String obtainStartTimestamp(SessionState sessionState) {
    Date date;
    if (sessionState.startTimestamp() == null) {
      date = new Date();
    } else {
      date = sessionState.startTimestamp();
    }
    return TelemetryUtils.generateCreateDateFormatted(date);
  }

  private static Location[] convertToArray(List<Location> locationList) {
    if (locationList == null) {
      return new Location[0];
    }
    return locationList.toArray(new Location[0]);
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/metrics/NavigationEventFactory.kt
package com.mapbox.services.android.navigation.v5.internal.navigation.metrics

import android.location.Location
import com.mapbox.android.telemetry.TelemetryUtils
import com.mapbox.services.android.navigation.v5.internal.navigation.routeprogress.MetricsRouteProgress
import com.mapbox.services.android.navigation.v5.utils.DistanceFormatter
import com.mapbox.services.android.navigation.v5.utils.extensions.ifNonNull
import java.util.Date

internal object NavigationEventFactory {

    internal const val EVENT_VERSION = 7

    fun buildNavigationDepartEvent(
        phoneState: PhoneState,
        sessionState: SessionState,
        metricProgress: MetricsRouteProgress,
        location: Location,
        sdkIdentifier: String
    ): NavigationDepartEvent = NavigationDepartEvent(phoneState).apply {
        setEvent(sessionState, metricProgress, location, sdkIdentifier, this)
    }

    fun buildNavigationCancelEvent(
        phoneState: PhoneState,
        sessionState: SessionState,
        metricProgress: MetricsRouteProgress,
        location: Location,
        sdkIdentifier: String
    ): NavigationCancelEvent = NavigationCancelEvent(phoneState).apply {
        setEvent(sessionState, metricProgress, location, sdkIdentifier, this)
        arrivalTimestamp = obtainArriveTimestamp(sessionState)
    }

    fun buildNavigationArriveEvent(
        phoneState: PhoneState,
        sessionState: SessionState,
        metricProgress: MetricsRouteProgress,
        location: Location,
        sdkIdentifier: String
    ): NavigationArriveEvent = NavigationArriveEvent(phoneState).apply {
        setEvent(sessionState, metricProgress, location, sdkIdentifier, this)
    }

    fun buildNavigationRerouteEvent(
        phoneState: PhoneState,
        sessionState: SessionState,
        metricProgress: MetricsRouteProgress,
        location: Location,
        sdkIdentifier: String,
        rerouteEvent: RerouteEvent
    ): NavigationRerouteEvent =
        NavigationRerouteEvent(phoneState, rerouteEvent, metricProgress).apply {
            setEvent(sessionState, metricProgress, location, sdkIdentifier, this)
            locationsBefore = convertToArray(sessionState.beforeEventLocations)
            locationsAfter = convertToArray(sessionState.afterEventLocations)
            secondsSinceLastReroute = sessionState.secondsSinceLastReroute
        }

    fun buildNavigationFeedbackEvent(
        phoneState: PhoneState,
        sessionState: SessionState,
        metricProgress: MetricsRouteProgress,
        location: Location,
        sdkIdentifier: String,
        feedbackDescription: String?,
        type: String?,
        feedbackScreenshot: String?,
        feedbackSource: String?
    ): NavigationFeedbackEvent =
        NavigationFeedbackEvent(phoneState, metricProgress).apply {
            setEvent(sessionState, metricProgress, location, sdkIdentifier, this)
            locationsBefore = convertToArray(sessionState.beforeEventLocations)
            locationsAfter = convertToArray(sessionState.afterEventLocations)
            description = feedbackDescription
            feedbackType = type
            screenshot = feedbackScreenshot
            source = feedbackSource
        }

    private fun setEvent(
        sessionState: SessionState,
        metricProgress: MetricsRouteProgress,
        location: Location,
        sdkIdentifier: String,
        navigationEvent: NavigationEvent
    ) {
        navigationEvent.absoluteDistanceToDestination =
            DistanceFormatter.calculateAbsoluteDistance(location, metricProgress)
        navigationEvent.distanceCompleted =
            (sessionState.eventRouteDistanceCompleted + metricProgress.distanceTraveled).toInt()
        navigationEvent.distanceRemaining = metricProgress.distanceRemaining
        navigationEvent.durationRemaining = metricProgress.durationRemaining
        navigationEvent.profile = metricProgress.directionsRouteProfile
        navigationEvent.legIndex = metricProgress.legIndex
        navigationEvent.legCount = metricProgress.legCount
        navigationEvent.stepIndex = metricProgress.stepIndex
        navigationEvent.stepCount = metricProgress.stepCount
        navigationEvent.estimatedDistance = metricProgress.directionsRouteDistance
        navigationEvent.estimatedDuration = metricProgress.directionsRouteDuration
        navigationEvent.startTimestamp = obtainStartTimestamp(sessionState)
        navigationEvent.eventVersion = EVENT_VERSION
        navigationEvent.sdkIdentifier = sdkIdentifier
        navigationEvent.sessionIdentifier = sessionState.sessionIdentifier
        navigationEvent.lat = location.latitude
        navigationEvent.lng = location.longitude
        navigationEvent.geometry = sessionState.currentGeometry()
        navigationEvent.simulation = sessionState.mockLocation
        navigationEvent.locationEngine = sessionState.locationEngineName
        navigationEvent.tripIdentifier = sessionState.tripIdentifier
        navigationEvent.rerouteCount = sessionState.rerouteCount
        navigationEvent.originalRequestIdentifier = sessionState.originalRequestIdentifier
        navigationEvent.requestIdentifier = sessionState.requestIdentifier
        navigationEvent.originalGeometry = sessionState.originalGeometry()
        navigationEvent.originalEstimatedDistance = sessionState.originalDistance()
        navigationEvent.originalEstimatedDuration = sessionState.originalDuration()
        navigationEvent.originalStepCount = sessionState.originalStepCount()
        navigationEvent.percentTimeInForeground = sessionState.percentInForeground
        navigationEvent.percentTimeInPortrait = sessionState.percentInPortrait
        navigationEvent.totalStepCount = sessionState.currentStepCount()
    }

    private fun obtainStartTimestamp(sessionState: SessionState): String {
        val date: Date = ifNonNull(sessionState.startTimestamp) {
            sessionState.startTimestamp
        } ?: Date()
        return TelemetryUtils.generateCreateDateFormatted(date)
    }

    private fun obtainArriveTimestamp(sessionState: SessionState): String {
        val date: Date = ifNonNull(sessionState.arrivalTimestamp) {
            sessionState.arrivalTimestamp
        } ?: Date()
        return TelemetryUtils.generateCreateDateFormatted(date)
    }

    private fun convertToArray(locationList: List<Location>?): Array<Location> =
        ifNonNull(locationList) {
            it.toTypedArray()
        } ?: arrayOf()
}
